staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deviceattribute
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	resourceapi "k8s.io/dynamic-resource-allocation/api"
+)
+
+// GetPCIeRootAttributeByPCIBusID retrieves the PCIe Root Complex for a given PCI Bus ID.
+// in BDF (Bus-Device-Function) format, e.g., "0123:45:1e.7".
+//
+// It returns a DeviceAttribute with the PCIe Root Complex information("pci<domain>:<bus>")
+// as a string value or an error if the PCI Bus ID is invalid or the root complex cannot be determined.
+//
+// ref: https://wiki.xenproject.org/wiki/Bus:Device.Function_(BDF)_Notation
+func GetPCIeRootAttributeByPCIBusID(pciBusID string) (DeviceAttribute, error) {
+	if pciBusID == "" {
+		return DeviceAttribute{}, fmt.Errorf("PCI Bus ID cannot be empty")
+	}
+
+	bdfRegexp := regexp.MustCompile(`^([0-9a-f]{4}):([0-9a-f]{2}):([0-9a-f]{2})\.([0-9a-f]{1})$`)
+	if !bdfRegexp.MatchString(pciBusID) {
+		return DeviceAttribute{}, fmt.Errorf("invalid PCI Bus ID format: %s", pciBusID)
+	}
+
+	// e.g. /sys/devices/pci0000:01/...<intermediate PCI devices>.../0000:00:1f.0,
+	sysDevicesPath, err := resolveSysDevicesPath(pciBusID)
+	if err != nil {
+		return DeviceAttribute{}, fmt.Errorf("failed to resolve sysfs path for PCI Bus ID %s: %w", pciBusID, err)
+	}
+
+	pciRootPart := strings.Split(strings.TrimPrefix(sysDevicesPath, sysfs.devices("")+"/"), "/")[0]
```

Comment:
Does it make sense to ensure there are at least some number of parts to the split? Or is it expected to ever have just the root by the time we reach this stage?

Reply:
> Does it make sense to ensure there are at least some number of parts to the split?
> Or is it expected to ever have just the root by the time we reach this stage?

Yes, I think so. 
`sysDevicesPath` is always of the form of `/sys/devices/pci0000:17/...`
ref: https://docs.kernel.org/PCI/sysfs-pci.html

Reply:
New method `resolvePCIeRoot` method was introduced in https://github.com/kubernetes/kubernetes/pull/132296/commits/a4fb1562f4a50b1709ff0cf6263fc12f99a6d3e8

extracting pcieRoot part is performed just after sysDevicesPath validation. So, I think it's more readable now:
https://github.com/kubernetes/kubernetes/blob/872f0682d585c4929ae2b278d03fa4ab27e30892/staging/src/k8s.io/dynamic-resource-allocation/deviceattribute/pci_linux.go#L89-L101